raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -180,6 +353,251 @@ public void testVoteRequestTimeout() throws Exception {
+
+    @Test
+    public void testHandleVoteRequestAsFollowerWithElectedLeader() throws Exception {
+        int leaderEpoch = 2;
+        int otherNodeId = 1;
+        int electedLeaderId = 3;
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId, electedLeaderId);
+
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withElectedLeader(leaderEpoch, electedLeaderId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertFalse(client.quorumState().followerStateOrThrow().hasVoted());
+
+        handleVoteRequest(leaderEpoch - 1, otherNodeId, client, false);
+
+        assertTrue(client.quorumState().isFollower());
+        assertFalse(client.quorumState().followerStateOrThrow().hasVoted());
+        assertEquals(client.quorumState().followerStateOrThrow().leaderId(), electedLeaderId);
+    }
+
+    @Test
+    public void testHandleVoteRequestAsFollowerWithVotedCandidate() throws Exception {
+        int leaderEpoch = 2;
+        int otherNodeId = 1;
+        int votedCandidateId = 3;
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId, votedCandidateId);
+
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withVotedCandidate(leaderEpoch, votedCandidateId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertTrue(client.quorumState().followerStateOrThrow().hasVoted());
+
+        handleVoteRequest(leaderEpoch - 1, otherNodeId, client, false);
+
+        assertTrue(client.quorumState().isFollower());
+        assertTrue(client.quorumState().followerStateOrThrow().hasVoted());
+        assertEquals(client.quorumState().followerStateOrThrow().votedId(), votedCandidateId);
+    }
+
+    private void handleVoteRequest(int lastEpoch,
+                                   int otherNodeId,
+                                   KafkaRaftClient client,
+                                   boolean voteGranted) throws Exception {
+        assertTrue(client.quorumState().isFollower());
+
+        VoteRequestData voteRequest = new VoteRequestData()
+                                          .setLastEpoch(lastEpoch)
+                                          .setLastEpochEndOffset(1)
+                                          .setCandidateId(otherNodeId)
+                                          .setCandidateEpoch(lastEpoch + 1);
+        channel.mockReceive(new RaftRequest.Inbound(
+            channel.newCorrelationId(), voteRequest, time.milliseconds()));
+
+        client.poll();
+
+        List<RaftMessage> unsent = channel.drainSendQueue();
+
+        assertTrue(unsent.size() >= 1);
+        RaftMessage lastMessage = unsent.get(unsent.size() - 1);
```

Comment:
Similar to previous comment. If the ordering is not semantically significant, then we should not have to rely on it in the test case.